include/fmt/core.h

```diff
@@ -2244,21 +2255,87 @@ class compile_parse_context
   using base::check_arg_id;
 };
 
+template <typename T> constexpr bool check_named_argument() {
+  return !detail::is_named_arg<T>() ||
+         (detail::is_named_arg<T>() && detail::is_statically_named_arg<T>());
+}
+
+template <typename... Args, FMT_ENABLE_IF(sizeof...(Args) == 0)>
+constexpr bool check_named_arguments() {
+  return true;
+}
+
+template <typename T, typename... Args> constexpr bool check_named_arguments() {
+  return check_named_argument<T>() && check_named_arguments<Args...>();
+}
+
+constexpr int invalid_arg_index = -1;
+
+#if FMT_USE_NONTYPE_TEMPLATE_PARAMETERS
+template <int N, typename T, typename... Args, typename Char>
+constexpr int get_arg_index_by_name(basic_string_view<Char> name) {
+  if constexpr (detail::is_statically_named_arg<T>()) {
+    if (name == T::name) return N;
+  }
+  if constexpr (sizeof...(Args) == 0) {
+    return invalid_arg_index;
+  } else {
+    return get_arg_index_by_name<N + 1, Args...>(name);
+  }
+}
+
+template <typename... Args, typename Char>
+constexpr int get_arg_index_by_name(basic_string_view<Char> name) {
+  if constexpr (sizeof...(Args) == 0) {
+    return invalid_arg_index;
+  } else {
+    return get_arg_index_by_name<0, Args...>(name);
+  }
+}
+#else
+template <typename... Args, typename Char>
+constexpr int get_arg_index_by_name(basic_string_view<Char>) {
+  return invalid_arg_index;
+}
+#endif
+
 template <typename Char, typename ErrorHandler, typename... Args>
 class format_string_checker {
  public:
   explicit FMT_CONSTEXPR format_string_checker(
       basic_string_view<Char> format_str, ErrorHandler eh)
       : context_(format_str, num_args, eh),
-        parse_funcs_{&parse_format_specs<Args, parse_context_type>...} {}
+        parse_funcs_{&parse_format_specs<Args, parse_context_type>...} {
+    static_assert(
+        check_named_arguments<Args...>(),
+#if FMT_USE_NONTYPE_TEMPLATE_PARAMETERS
+        "use statically named arguments with compile-time format strings: "
+        "fmt::arg(\"name\", value) -> \"name\"_a = value"
+#else
+        "compile-time checks don't support named arguments, "
+        "please use C++20 or/and a newer compiler"
+#endif
+    );
```

Comment:
I prefer to leave this check even if it has some small effect on use without named arguments because the error messages without this check are... bad, and it's not just one compiler:
* errors with GCC: https://godbolt.org/z/e3YsvhvGW
* errors with Clang: https://godbolt.org/z/ajaEP74z9 (NTTP enabled explicitly for C++20 here)
* errors with MSVC: https://godbolt.org/z/nGKjPEa8G - I can't find an error without this check here

Reply:
Named arguments are uncommon, it's OK if error message for invalid use is not great.

Reply:
👎 - this is how I express my disagreement 😉

For me, named arguments are much readable and maintainable than just several `{}` because it's a big pain to edit those usages considering that I have to maintain the correct order. While with named arguments, I can add another `ARG(var)` and call it a day. Especially if I use `FMT_COMPILE`, then it's as fast as just replacement fields.

But ok, I will remove this check reluctantly 😔

Reply:
Thinking more of it, we shouldn't give an error at all, good or bad, but instead fallback on runtime formatting as it is done in other cases.

Reply:
But the point of `FMT_STRING` and `format_string_checker` is to notify that there is an error at compile-time, isn't it? And since they cannot check that case does not mean that they should skip the check.

Reply:
Only if such check if possible, otherwise `FMT_STRING` is a noop. So for `fmt::arg` it should always be a noop unless `FMT_ENFORCE_COMPILE_STRING` is defined.

Reply:
But that's not how `FMT_STRING` and `FMT_ENFORCE_COMPILE_STRING` relates to each other right now:

> Compile-time checks are enabled when using FMT_STRING. ...
> To force the use of compile-time checks, define the preprocessor variable FMT_ENFORCE_COMPILE_STRING. When set, functions accepting FMT_STRING will fail to compile with regular strings. Runtime-checked formatting is still possible using fmt::vformat, fmt::vprint, etc.

Reply:
You are right about `FMT_ENFORCE_COMPILE_STRING`. The point about `FMT_STRING` still holds though and it will become even more important to make `fmt::arg` not error once we enable compile-time checks by default. Otherwise it will become unusable.

Reply:
IMHO, it will probably be OK to have this runtime fallback for compile-time checks which are invoked implicitly, like `_format` UDL one or the argument-casting one. But if `FMT_STRING` macro is still available after implementing the latter check, then it should fail compilation because, in this case, the compile-time check is requested explicitly. Maybe it's just my pain again, but I want this explicit check.